### ISSUE TITLE
Docs: Not all reported problems of `no-redeclare` are covered by TypeScript

### ISSUE THREAD

**Issue by ulrichstark** [typescript-eslint/typescript-eslint#11461]
### Before You File a Documentation Request Please Confirm You Have Done The Following...

- [x] I have looked for existing [open or closed documentation requests](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Adocumentation) that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Suggested Changes

Adjust the "Danger" banner of the `no-redeclare` lint rule that currently states:

> The code problem checked by this ESLint rule is automatically checked by the TypeScript compiler. Thus, it is not recommended to turn on this rule in new TypeScript projects. You only need to enable this rule if you prefer the ESLint error messages over the TypeScript compiler error messages.

That's incorrect, because this rule does report naming collisions between a const and a type, while TypeScript doesn't. This rule uncovered an accidental issue of an incorrect name in our code we wouldn't have discovered with TypeScript. See reproduction link below.

I'm not sure what a better message would be or if that means the banner needs to removed from the docs page.

### Affected URL(s)

https://typescript-eslint.io/rules/no-redeclare/

### Additional Info

https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=JYWwDg9gTgLgBALzgMyhEcBECIBNMDcAUEQKYAekscAxhAHYDO8AYhBAMo0AWpIAhnAC8iAHQQARgCtSNGAAoA3nDBowpWAE8AXGOZRg9AObyAlHAC%2Bp4mUrR4MTerhtOPPoJEJRh5BoA8juoQyC7sXLwCAHzEQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaaRAE0QGN4BDE9KfAM0cgBpxtIABAFwE8ADigrQAlgJ6EU8Ubh4B6AsTKUadDJETRo%2BaJA4BfEAaA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

**Comment by bradzacher** [comment]
This usecase is actually the entire reason that we have marked the rule with a big banner.

----

First of all - this isn't actually a redeclaration - adding this functionality to the rule was an abuse of the term.

A redeclaration occurs when you redeclare a variable within the same scope. I.e

```ts
var x = 1;
var x = 2;
```

In a simplified nutshell types and values exist in distinct scopes. So you cannot redeclare a value as a type or vice versa.

This is an important distinction because a redeclaration is usually a bug because you physically lose the old value!
However declaring a name as a value and a type loses no information ever - TS is fully able to access both instances in the correct usecases.

---

Additionally we have found that the vast majority of people either do not care about this error and in fact many people find it actively harmful to report on it.

In fact are many popular libraries (such as zod) which are actively designed around the practice!

---

IMO this is all working as intended. Thoughts @typescript-eslint/triage-team?

**Comment by ulrichstark** [comment]
Thanks for the explanation! Your reasoning makes sense.

Then what do you think about adding a new rule that catches name collisions between type and value namespace?
Because today I learnt that I can `import { FooSchema } from "./module-of-my-link-above"` and even though it's only one import, this imports both the type and the value and uses one of both depending on the context. That's pretty confusing in my opinion.

**Comment by bradzacher** [comment]
> Because today I learnt that I can import { FooSchema } from "./module-of-my-link-above" and even though it's only one import, this imports both the type and the value and uses one of both depending on the context.

`enum`, `class` and `namespace` all already behave this way, btw!
All of these are "dual type/value" constructs!

```ts
enum Foo { A = 1 }

const test1 = Foo.A;
type Test2 = Foo.A;

const test3 = Foo;
const Test4 = Foo;

const test5 = typeof Foo;
const Test6 = typeof Foo;
```

```ts
class Bar { static prop1 = 1; prop2 = 2 }

const test7 = Bar;
type Test8 = Bar;

const test9 = typeof Bar;
type Test10 = typeof Bar;

const test11 = Bar['prop1'];
type Test12 = Bar['prop2'];
```

```ts
namespace Baz {
    export const x = 1;
    export type T = 2;
}

const test13 = Baz.x;
type Test14 = Baz.T;
```
[playground](https://www.typescriptlang.org/play/?#code/KYOwrgtgBAYg9nKBvKBBKBeKBGKBfAWAChiBjOEAZwBcprgbct44A6VAbmOoE8AHYFAAqDagCZMsBOy4ki5KrXo0AzJJayFNYaIAs6hLLIVty6gFZJvAXABmUuJpO0RNAGxX+wOw6NyyADYAhpSUUABCQQBOyFA0QdQAlqRQfFFwfEw4HKnpfBJYEoRyWkqiAOySkVGy1oKu1AAcVdF+pXSiAJyeNvbVtV46jAAMPd59rcTGih2MWdUA2gDkaRnYSwC6AwJD1NgFEdHLq-mbflNEIEEQDHxBpIKRAF7IxFDvUMAAHnxwUbTtL6SbCyD6fH5-JSDISSMSyYrTUyibBqLDPVhfbb1ZH6NFBJ6sISyIA)

So being able to manually declare a name as a "dual type/value" construct isn't unusual or unexpected!

**Comment by kirkwaiblinger** [comment]
> So being able to manually declare a name as a "dual type/value" construct isn't unusual or unexpected!

The current state of [the rule docs](https://typescript-eslint.io/rules/no-redeclare/#ignoredeclarationmerge) imply the opposite perspective.

> Even with this option set to true, this rule will report if you name a type and a variable the same name. This is intentional. Declaring a variable and a type the same is usually an accident, and it can lead to hard-to-understand code. If you have a rare case where you're intentionally naming a type the same name as a variable, use a disable comment

Overall, I think it's a bit contradictory to have the banner at the top saying "you don't need this rule because it's checked by TS" on an extension rule which unconditionally adds a TS-specific check that is not checked by TS. Might be more of a docs issue than a functionality issue, but probably something worth cleaning up to be consistent.

---

I tend to agree that types and values of the same name are unexpected and probably unintentional outside of advanced users / library authors. But, they also don't have much potential for harm that I can think of, and are probably not encountered unintentionally terribly often since people tend to `lowerCamelCase` names and `UpperCamelCase` types. So 🤷 whether this check is important and should be enabled by default or not, since the users who do declaration-merge types and values may well be the advanced users doing so intentionally.

**Comment by JoshuaKGoldberg** [comment]
> since the users who do declaration-merge types and values may well be the advanced users doing so intentionally.

I think this is true in isolation, but the problem arises when newcomers see and copy this practice. The recommended Zod practice in particular is a big one. Example from https://zod.dev/basics?id=inferring-types:

```ts
const Player = z.object({ 
  username: z.string(),
  xp: z.number()
});
 
// extract the inferred type
type Player = z.infer<typeof Player>;
 
// use it in your code
const player: Player = { username: "billie", xp: 100 };
```

I've seen quite a few developers get tripped up by this. It's really unfortunate. 

Speaking of which, we should really file a discussion/issue on Zod to start a dialog...

**Comment by kirkwaiblinger** [comment]
@JoshuaKGoldberg What is the harm with that snippet? Just surprise? I guess I personally would write 
```ts
const playerSchema = ...; 
type Player = z.output<typeof Player>;
const player: Player = ...
```

**Comment by JoshuaKGoldberg** [comment]
For people who don't yet really understand types vs. values, it's _really_ confusing and off-putting to see one name refer to both. In my experiences teaching new folks -especially those coming from other runtime-oriented typed languages like C# and Java- the _"wait, what's erased?"_ question has been a pretty consistent confusion point.

**Comment by bradzacher** [comment]
Most people will see the docs and do what the docs tell them!
[A quick query of public code shows a bit of a mixed bag](https://sourcegraph.com/search?q=context:global+/type+%5BA-Z%5D%5Cw%2B+%3D+z%5C.infer/+AND+/const+%5BA-Z%5D%5Cw%2B+%3D+z%5C./&patternType=keyword&case=yes&sm=0)

Some people do the shadowing, other people do `const FooSchema` + `type Foo`.

**Comment by kirkwaiblinger** [comment]
So is a behavior change warranted, too, to un-disable the base rule in the `@typescript-eslint/eslint-recommended` config ([currently it's there](https://github.com/typescript-eslint/typescript-eslint/blob/69e2f6c0d371f304c6793ba1801adde10a89372b/packages/eslint-plugin/src/configs/eslint-recommended-raw.ts#L34)), but enable the extension rule in a recommended config? Or do we just remove the danger banner and leave it at that? Or do we want to also plan to split out the declaration-merge-y stuff from this rule?

I'm thinking remove banner for now. Consider breaking behavior separately?

**Comment by Josh-Cena** [comment]
We should note that it _overlaps_ with TS, but—as I continuously say—there are many legit cases of type-value shadowing and pedagogical concerns aside there really aren't many bugs that could be caused by this because they live in entirely different spaces. It's like having `class Foo { prop; static prop; }`. But I'm not comfortable with simply removing the banner.

**Comment by bradzacher** [comment]
> So is a behavior change warranted, too, to un-disable the base rule in the @typescript-eslint/eslint-recommended config ([currently it's there](https://github.com/typescript-eslint/typescript-eslint/blob/69e2f6c0d371f304c6793ba1801adde10a89372b/packages/eslint-plugin/src/configs/eslint-recommended-raw.ts#L34)), but enable the extension rule in a recommended config?

IMO - no.
Our recommended set must be unopinionated and about catching bugs.
The base `no-redeclare` or our extension version of it do not catch bugs in addition to what TS already reports on.

To add it to our recommended set would be setting a stance and banning the practice of type/value shadowing. We did this in the past and we very intentionally removed it _because it is an opinionated, stylistic decision_.

---

It may be slightly inaccurate to just say "no-redeclare reports on the same thing TS reports on" because it does also prevent type/value shadowing. But as we've established that whilst this might be surprising behaviour - it is never going to be a bug. So IMO the banner is "correct enough" for most people and serves its purpose of warding people off a rule which will primarily be a noisy burden for them.

My 2c: close and move on as the docs and configs are correct for the vast majority of users.

**Comment by JoshuaKGoldberg** [comment]
> My 2c: close and move on as the docs and configs are correct for the vast majority of users.

👍 from me, except I do think the docs could be made a little clearer. They do mention under https://typescript-eslint.io/rules/no-redeclare/#ignoredeclarationmerge that reporting on a variable with a type and variable is intentional. But they describe it as a "rare case" even though some libraries actually recommend it.

Proposal: how about we add a `###` for an FAQ / explainer summarizing this issue?

**Comment by Josh-Cena** [comment]
IMO the current docs are contradictory in a fresh pair of eyes and some revision is necessary. The banner strongly suggests that the errors are equivalent because it goes as far as saying "You only need to enable this rule if you prefer the ESLint error messages over the TypeScript compiler error messages." which suggests the difference is cosmetic. But right below we provide a configuration that's clearly not 1-1 mapped with TS behavior, and the default behavior of this rule is also not 1-1 mapped with the TS messages. So the banner's message, which is that there's 0 difference, does not hold, and we need to change its wording.
